File: DSA Unit 05/Lec 34/Q1.cpp
Single File Programming Question
Problem Statement



Mitchell Starc  is implementing a hash table using double hashing for collision resolution. He needs to insert a key into the table and handle collisions using a second hash function. If a collision occurs, double hashing will be used to find the next available position. After inserting the key, Mitchell Starc needs to know the position where the key was placed.



Help him implement the task.



The first hash function is: h1(key) = key % TABLE_SIZE.

The second hash function is: h2(key) = 1 + (key % (TABLE_SIZE - 1)).

The position is calculated as: position = (h1(key) + i * h2(key)) % TABLE_SIZE,

where i is the number of collisions that have occurred.

Input format :
The first line of input consists of an integer key, representing the value to be inserted into the hash table.

Output format :
The output should print: The key <key> is inserted at position <position>.,

where <key> is the key inserted and <position> is the position in the hash table where the key was placed.



Refer to the sample output for formatting specifications.

Code constraints :
The given test cases fall under the following constraints:

TABLE_SIZE = 7.

The key is an integer.

Sample test cases :
Input 1 :
42
Output 1 :
The key 42 is inserted at position 0.
Input 2 :
59
Output 2 :
The key 59 is inserted at position 3



#include <bits/stdc++.h>
using namespace std;

int main() {
    const int TABLE_SIZE = 7;

    int key;
    cin >> key;

    int h1 = ((key % TABLE_SIZE) + TABLE_SIZE) % TABLE_SIZE;
    int h2 = 1 + ((key % (TABLE_SIZE - 1)) + (TABLE_SIZE - 1)) % (TABLE_SIZE - 1);

    int i = 0;
    int position = (h1 + i * h2) % TABLE_SIZE;

    cout << "The key " << key << " is inserted at position " << position << ".";
    return 0;
}

File: Java/Unit 03/Lec 18/Q1.java
Single File Programming Question
Problem Statement



Imagine Tim is developing a user authentication system for a secure application. 



As part of the security checks, users need to enter two passphrases and the system VowelCountMatching class must verify if two entered passphrases have the same number of vowels. Provide the result as 'Access Granted' if true, or 'Access Denied' otherwise, along with the count of Vowels using the equals() method.



Tim needs your assistance in completing the system. Help him by writing the code.

Input format :
The input consists of two strings, each entered on a separate line, representing the two passphrases provided by the user.

Output format :
The first line of output prints “Access Granted” if both passphrases contain the same number of vowels; otherwise, it prints “Access Denied”.

The second line prints “First String: <count>”, where <count> is the number of vowels in the first passphrase.

The third line prints “Second String: <count>”, where <count> is the number of vowels in the second passphrase.



Refer to the sample output for formatting specifications.

Code constraints :
In this scenario, the test cases fall under the following constraints:

1 ≤ length of the input strings ≤ 100 characters

Sample test cases :
Input 1 :
Username
Password
Output 1 :
Access Denied
First String: 4
Second String: 2
Input 2 :
AEIOU
aeiou
Output 2 :
Access Granted
First String: 5
Second String: 5

Whitelist
Set 1:
equals

import java.util.Scanner;

// Class to match vowel counts
class VowelCountMatching {

    String text;

    // Constructor
    VowelCountMatching(String text) {
        this.text = text;
    }

    // Method to count vowels in a string
    int countVowels(String str) {

        int count = 0;
        str = str.toLowerCase();

        for (int i = 0; i < str.length(); i++) {

            char ch = str.charAt(i);

            if (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u') {
                count++;
            }
        }

        return count;
    }

    // Override equals method to compare vowel counts
    @Override
    public boolean equals(Object obj) {

        if (!(obj instanceof VowelCountMatching)) {
            return false;
        }

        VowelCountMatching other = (VowelCountMatching) obj;

        int count1 = countVowels(this.text);
        int count2 = countVowels(other.text);

        return count1 == count2;
    }
}

// Main class
public class Main {

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);

        String firstString = scanner.nextLine();
        String secondString = scanner.nextLine();

        VowelCountMatching object1 = new VowelCountMatching(firstString);
        VowelCountMatching object2 = new VowelCountMatching(secondString);

        boolean result = object1.equals(object2);

        System.out.println(result ? "Access Granted" : "Access Denied");
        System.out.println("First String: " + object1.countVowels(firstString));
        System.out.println("Second String: " + object2.countVowels(secondString));

        scanner.close();
    }
}
